File: frontend/src/api/quiz.ts
import type { ApiResponse } from "../types/auth";
import type { QuizQuestion, QuizSession, RecommendationResult } from "../types/quiz";
import api from "./client";

export async function startQuiz(language: "en" | "zh"): Promise<QuizSession> {
  const { data } = await api.post<ApiResponse<QuizSession>>("/quiz/sessions", { language });
  return data.data;
}

export async function getQuizQuestions(): Promise<QuizQuestion[]> {
  const { data } = await api.get<ApiResponse<QuizQuestion[]>>("/quiz/questions");
  return data.data;
}

export async function saveQuizAnswers(
  sessionToken: string,
  answers: Record<string, string | string[]>,
): Promise<QuizSession> {
  const { data } = await api.post<ApiResponse<QuizSession>>(
    `/quiz/sessions/${encodeURIComponent(sessionToken)}/answers`,
    { answers },
  );
  return data.data;
}

export async function getRecommendations(sessionToken: string): Promise<RecommendationResult> {
  const { data } = await api.post<ApiResponse<RecommendationResult>>(
    `/quiz/sessions/${encodeURIComponent(sessionToken)}/recommendations`,
  );
  return data.data;
}

File: backend/app/api/v1/quiz.py
from fastapi import APIRouter, HTTPException

from app.api.deps import DbSession
from app.schemas.common import SuccessResponse
from app.schemas.quiz import (
    QuizAnswersRequest,
    QuizQuestionResponse,
    QuizSessionCreate,
    QuizSessionResponse,
    RecommendationResponse,
)
from app.services.quiz import create_session, get_questions, get_session, recommend, save_answers


router = APIRouter()


@router.post("/quiz/sessions", response_model=SuccessResponse[QuizSessionResponse], status_code=201)
async def start_quiz(payload: QuizSessionCreate, db: DbSession):
    session = await create_session(db, payload.language)
    return SuccessResponse(
        data=QuizSessionResponse(
            session_token=session.session_token,
            language=session.language,
            status=session.status,
        )
    )


@router.get("/quiz/questions", response_model=SuccessResponse[list[QuizQuestionResponse]])
async def quiz_questions(db: DbSession):
    return SuccessResponse(data=await get_questions(db))


@router.post("/quiz/sessions/{session_token}/answers", response_model=SuccessResponse[QuizSessionResponse])
async def submit_answers(session_token: str, payload: QuizAnswersRequest, db: DbSession):
    session = await get_session(db, session_token)
    if session is None:
        raise HTTPException(status_code=404, detail={"code": "QUIZ_SESSION_NOT_FOUND", "message": "Quiz session not found"})
    session = await save_answers(db, session, payload.answers)
    return SuccessResponse(
        data=QuizSessionResponse(
            session_token=session.session_token,
            language=session.language,
            status=session.status,
        )
    )


@router.post(
    "/quiz/sessions/{session_token}/recommendations",
    response_model=SuccessResponse[RecommendationResponse],
)
async def recommendations(session_token: str, db: DbSession):
    session = await get_session(db, session_token)
    if session is None:
        raise HTTPException(status_code=404, detail={"code": "QUIZ_SESSION_NOT_FOUND", "message": "Quiz session not found"})
    return SuccessResponse(data=await recommend(db, session))
